File: postorder-binary-search/postorder-binary-search.py
def postorder(node):
	if node is None: return []

	left = postorder(node["left"])
	right = postorder(node["right"])

	return left + right + [node["value"]]

def postorder_with_accumulator(node, acc=None):
	if acc is None: acc = []
	if node is None: return acc

	postorder_with_accumulator(node["left"], acc)
	postorder_with_accumulator(node["right"], acc)
	acc.append(node["value"])

	return acc

tree = {
	"value": "A",
	"left": {
		"value": "B",
		"left": { "value": "D", "left": None, "right": None },
		"right": { "value": "E", "left": None, "right": None },
	},
	"right": {
		"value": "C",
		"left": None,
		"right": { "value": "F", "left": None, "right": None },
	},
}

print(postorder(tree))
print(postorder_with_accumulator(tree))

File: postorder-binary-search/postorder-binary-search.js
function postorder(node) {
	if (!node) return [];

	const left = postorder(node.left);
	const right = postorder(node.right);

	return [...left, ...right, node.value];
}

function postorderWithAccumulator(node, acc = []) {
	if (!node) return acc;

	postorderWithAccumulator(node.left, acc);
	postorderWithAccumulator(node.right, acc);
	acc.push(node.value);

	return acc;
}

function iterativePostorder(tree) {
	if (!tree) return [];

	const stack = [tree];
	const result = [];

	while (stack.length > 0) {
		const node = stack.pop();

		result.push(node.value);

		if (node.left) stack.push(node.left);
		if (node.right) stack.push(node.right);
	}

	return result.reverse();
}

const tree = {
	value: "A",
	left: {
		value: "B",
		left: { value: "D", left: null, right: null },
		right: { value: "E", left: null, right: null },
	},
	right: {
		value: "C",
		left: null,
		right: { value: "F", left: null, right: null },
	},
};

console.log(postorder(tree));
console.log(postorderWithAccumulator(tree));
console.log(iterativePostorder(tree));
